milli/src/search/criteria/proximity.rs

```diff
@@ -192,22 +192,42 @@ fn resolve_candidates<'t>(
                     let most_right = words
                         .last()
                         .map(|w| Query { prefix: false, kind: QueryKind::exact(w.clone()) });
-                    let mut candidates = None;
-                    for slice in words.windows(2) {
-                        let (left, right) = (&slice[0], &slice[1]);
-                        match ctx.word_pair_proximity_docids(left, right, 1)? {
-                            Some(pair_docids) => match candidates.as_mut() {
-                                Some(candidates) => *candidates &= pair_docids,
-                                None => candidates = Some(pair_docids),
-                            },
-                            None => {
-                                candidates = None;
+                    let mut candidates = RoaringBitmap::new();
+                    let mut first_iter = true;
+                    let winsize = words.len().min(7);
+
+                    for win in words.windows(winsize) {
+                        // Get all the documents with the matching distance for each word pairs.
+                        let mut bitmaps = Vec::with_capacity(winsize.pow(2));
+                        for (offset, s1) in win.iter().enumerate() {
+                            for (dist, s2) in win.iter().skip(offset + 1).enumerate() {
+                                match ctx.word_pair_proximity_docids(s1, s2, dist as u8 + 1)? {
+                                    Some(m) => bitmaps.push(m),
+                                    // If there are no document for this distance, there will be no
+                                    // results for the phrase query.
+                                    None => return Ok(Default::default()),
+                                }
+                            }
+                        }
+
+                        // We sort the bitmaps so that we perform the small intersections first, which is faster.
+                        bitmaps.sort_unstable_by(|a, b| a.len().cmp(&b.len()));
+
+                        for bitmap in bitmaps {
+                            if first_iter {
+                                candidates = bitmap;
+                                first_iter = false;
+                            } else {
+                                candidates &= bitmap;
+                            }
+                            // There will be no match, return early
+                            if candidates.is_empty() {
```

Comment:
This is a copy-paste of the other modified file, we didn't have the same behavior in proximity as in the main function, this difference was hiding the bug located in the main function.

Reply:
Should we maybe factor it out in a separate function if we want the two copies to stay the same? It's likely that they will diverge over time if we are not careful. 

As an alternative, add a comment linking the two copies together, something like:
```rust
// The code below is exactly the same as #hGVU7jMhU4Mwr4XoU77B
let mut candidates = RoaringBitmap::new();


// and for the second copy as well:
// The code below is exactly the same as #hGVU7jMhU4Mwr4XoU77B
let mut candidates = RoaringBitmap::new();

```

Reply:
> Should we maybe factor it out in a separate function if we want the two copies to stay the same?

That would be cool if it’s possible